### ISSUE TITLE
improper locking bugs due the unrelesed locks

### ISSUE THREAD

**Issue by ycaibb** [archiecobbs/s3backer#151]
Hi, developers, thank you for your checking! Should the lock be released before destroying it? According to [this,](https://pubs.opengroup.org/onlinepubs/007904875/functions/pthread_mutex_destroy.html) it says` It shall be safe to destroy an initialized mutex that is` **unlocked**.

https://github.com/archiecobbs/s3backer/blob/016c3fe69760b6a0b03193e218b19488fa0862e8/block_cache.c#L490

```
static void
block_cache_destroy(struct s3backer_store *const s3b)
{
    ...;

    pthread_mutex_lock(&priv->mutex); // 
    ...;
    pthread_mutex_destroy(&priv->mutex); // do not released the lock before destorying it
    free(priv);
    free(s3b);
}
```

**Comment by ycaibb** [comment]
Similar to the below the situation
https://github.com/archiecobbs/s3backer/blob/016c3fe69760b6a0b03193e218b19488fa0862e8/ec_protect.c#L290-L312

**Comment by ycaibb** [comment]
https://github.com/archiecobbs/s3backer/blob/016c3fe69760b6a0b03193e218b19488fa0862e8/zero_cache.c#L321-L338

**Comment by archiecobbs** [comment]
Thanks! Should be fixed in 9f90b24.
